docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,507 @@
+provisioning volumes based on *volume_name* passed in by the CO, in the
+CreateVolume request to the CS.
+
+As per the CSI spec, creation requests come with a *volume_name* that is
+controlled by the CO and is unique across different volume requests, and the
+same for a given volume to maintain idempotent create requests.
+
+The CSI plugin returns a *volume_id* that is passed to other plugin RPCs to
+refer to the volume, and is generated by the plugin. Snippet from the spec
+about the *volume_id*,
+
+```
+  // This field MUST contain enough information to uniquely identify
+  // this specific volume vs all other volumes supported by this plugin.
+  // This field SHALL be used by the CO in subsequent calls to refer to
+  // this volume.
+```
+
+#### Name terminology
+
+- **ImageName**: Name of the image as created on Ceph (rbd image name, or
+  CephFS sub-directory)
+- **volume_id**: Identifier for the *ImageName* as shared with the CO system
+- **csi-id**: Immutable CSI instance ID, to distinguish between different CSI
+  instances, within or across CO systems, that may use the same Ceph cluster
+- **volume_name**: Name of the volume as recognized by the CO system, passed in
+  only during VolumeCreate operation
+- **cluster-id**: Ceph cluster fsid
+- **pool-name**: Ceph pool on which image is created
+- **fsname**: Ceph FS filesystem name, on which volume is created
+
+#### Name interpolation rules
+
+1. *ImageName* is known only to Ceph, *volume_id* is known only to the CO system
+1. We should be able to generate *ImageName* given a *volume_id*
+  - This is needed in most NS operations, that receive only the *volume_id*
+    and conditionally the *volume_context* and *publish_context*
+  - *volume_context* and *publish_context* are only cached, and not
+    essentially persisted, hence we cannot rely on payload from the same
+    to complete the transformation
+  - This transformation is needed by the CS specifically for DeleteVolume,
+    DeleteSnapshot and ControllerExpandVolume operations
+1. We should generate the same *ImageName*, when a Create request comes in with
+  the same *volume_name*
+  - CreateVolume may retry with the same *volume_name* on timeouts of the
+    request, and hence we need to generate the same *ImageName* for a given
+    *volume_name*
+1. We should be able to distinguish *ImageName*s that come from an instance of
+  the CSI plugin
+  - When dealing with a common Ceph cluster shared across CO systems, we need
+    the *ImageName* to be distinguishable, to avoid collisions and also to
+    be able support/aid operations like ListVolumes/Snapshots
+  - The same also applies when dealing with different CSI plugin instance in
+    the same CO system, using the same Ceph cluster
+
+#### Using Ceph OMaps for idempotent RPCs
+
+To preserve the idempotent CreateVolume RPC requirement, across *ImageName* for
+a given *volume_name*, it is proposed to use a per-CSI instance Ceph omap named
```

Comment:
you mean Ceph object named `csi.volumes.[csi-id]` with omap keys `csi.volumes.[vol_name]` and values 'imagenames'

Reply:
If those are the right terms then yes, will modify the doc to word it so.